### PR TITLE
Support vue listing

### LINKED ISSUE #15: Listing Improvements
There's a couple of improvements I'd like to make to the index listing:

* Ability for filters, just like the entries listing
* Real-time search (over AJAX, instead of full page reloads)
* Confirmation on the delete action #13
* Ability to use [Actions](https://statamic.dev/extending/actions#content) on the Runway listing #14 

I still need to investigate how this would work behind the scenes, especially with the filters..

## How should it work?

Will probably need to convert the index listing from Blade to Vue so I can do the real-time-ness that's required for some of the features.

## Use cases

Where you want to use Runway as a replacement for entries.

### PR DESCRIPTION
**Pull request by ryanmitchell**
## Description

First pass at this, there will definitely be more work required!

This PR brings Vue-style to the listing pages, allowing for searching, scoping and saving of queries.

It doesnt bring across the listing buttons, which will need done. You can see where it needs done around line 68 of ListingView.vue

You'll need to generate the mix js again - I didn't know if you would want the committed or if thats a build process thing.

Closes #15

### CODE REVIEW

**Comment by duncanmcclean** [comment]
Wow, thanks! That's amazing. I'll take a look at this tonight.

I need to change out 'listing buttons' for proper actions which I'll do at some point, probably before this is released.

Don't worry about the frontend assets - they should all be ignored anyway and they're recompiled independently for every release.

**Comment by ryanmitchell** [comment]
No problem - its mostly a rehash of what I have done elsewhere. I'm here if you need anything so just shout.

File: resources/js/components/Listing/DeletesListingRow.js
Comment (duncanmcclean) [inline resources/js/components/Listing/DeletesListingRow.js:8]: We probably don't need this `computed` code if it's just empty.

File: resources/views/index.blade.php
Comment (duncanmcclean) [inline resources/views/index.blade.php:17]: I think there might be something here that you've forgotten to change 😅  I imagine it can just be something like `runway.{{ $resource->handle() }}`

File: resources/views/index.blade.php
Comment (duncanmcclean) [inline resources/views/index.blade.php:21]: Would you mind hooking these URLs in to use the `cp_route` method (if possible), I don't love having URLs hard coded in places. Just personal preference.

File: src/Http/Controllers/ResourceController.php
Comment (duncanmcclean) [inline src/Http/Controllers/ResourceController.php:51]: Instead of using the `request()` helper here, can we use `$request`? Again, another thing that is personal preference.

**Review by duncanmcclean — CHANGES_REQUESTED** [review]
Not had time to properly pull it down it yet but here's a couple initial things I noticed when going through the code. Thanks again for your hard work on this one!

File: resources/js/components/Listing/DeletesListingRow.js
Comment (ryanmitchell) [reply resources/js/components/Listing/DeletesListingRow.js]: Done!

File: resources/views/index.blade.php
Comment (ryanmitchell) [reply resources/views/index.blade.php]: Yeah thats what I get for trying to be quick :) Changed!

File: resources/views/index.blade.php
Comment (ryanmitchell) [reply resources/views/index.blade.php]: Done - let me know if I've understood correctly.

File: src/Http/Controllers/ResourceController.php
Comment (ryanmitchell) [reply src/Http/Controllers/ResourceController.php]: Yep! Done.

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:64]: Not sure where the code for this is but just popping this here... I'm not sure why but it seems like the text isn't showing for any of these columns? Is this something my side? 

image

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:100]: Also not quite sure where to put this but I seem to get a pagination related page when I load up the listing page. Any ideas? 
image

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:66]: We seem to be taking the `editUrl` passed from the `listingConfig` and just adding to it here. Is there any way we can generate this for each row rather than doing it this way?

**Review by duncanmcclean — CHANGES_REQUESTED** [review]
Looks good, just had a proper dig into this PR. I've made a couple of changes, nothing personal, just some general tidy up:

* Moved the `listingConfig` prop from being generated in the view, instead it's now passed in from the controller.

* I've moved the `ResourceCollection` to `Http/Resources` to keep it inline with the standard folder structure for resources in a Laravel app.

* I've also moved the Listing API endpoint into its own controller and changed the route slightly. 

There's a couple of issues/things I'm not quite sure how to fix. Hopefully you'll be able to help.

**Comment by duncanmcclean** [comment]
Also, just noticed the 'displayed columns' selection doesn't seem to save properly. I removed a column, hit save, waited a couple secs, refreshed and the column is still there...

**Comment by ryanmitchell** [comment]
That should be column preferences saving now

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:68]: When I click the 'Edit' button in the dropdown, I get redirected to `/cp/NaN`. However, if I look at the `ListingView` dev tools I can see `editUrl` is set properly so not sure where it's breaking..

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:71]: Not sure why but I'm unable to see the 'Delete' dropdown here.. 

image

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [reply resources/js/components/Listing/ListingView.vue:93]: Another thing here - I don't see a pagination bar at the bottom of the `ListingView` which I should have because I have several hundred items for that model.

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [inline resources/js/components/Listing/ListingView.vue:53]: Is there any way we can make the first column a link to the 'Edit' page? Currently the only way for me seems to be opening the dropdown.

I'm guessing this is something to do with `has_link` in the `ResourceListingController`

**Review by duncanmcclean — CHANGES_REQUESTED** [review]
Thanks! The filters are saving properly now which is great. I just have a couple of other bits and bobs that need fixed up before I can merge (I'm really excited to merge this).

File: resources/js/components/Listing/ListingView.vue
Comment (ryanmitchell) [reply resources/js/components/Listing/ListingView.vue:68]: Hmm, definitely working ok for me. Did you build from the latest ListingVue i updated?

File: resources/js/components/Listing/ListingView.vue
Comment (ryanmitchell) [reply resources/js/components/Listing/ListingView.vue:71]: Should be there in the next push.

File: resources/js/components/Listing/ListingView.vue
Comment (ryanmitchell) [reply resources/js/components/Listing/ListingView.vue:53]: First column Edit is done.
Pagination - definitely there for me - try messing with line 38 in ResourceListingController?

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [reply resources/js/components/Listing/ListingView.vue:74]: Working fine now, thanks!

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [reply resources/js/components/Listing/ListingView.vue:77]: Whoop whoop, the button shows. Although, there's a really weird behaviour when I try to use it. Loom: https://www.loom.com/share/727e507d35e04a5abed164f292358fd1

File: resources/js/components/Listing/ListingView.vue
Comment (duncanmcclean) [reply resources/js/components/Listing/ListingView.vue:53]: Thanks! Both things seem to work now.

**Comment by duncanmcclean** [comment]
Thanks again! Just one more thing...

**Comment by ryanmitchell** [comment]
Fixed up the crazy number of overlays issue… the columns still disappear when I click on delete which is odd, as I doesn’t on the code I based this on. Makes me think something has changed in the statamic Listing.vue we inherit from.

If i get some time tomorrow I’ll try and investigate, but may not.

**Comment by duncanmcclean** [comment]
That's fine if not, I can take a look. 👍

**Comment by duncanmcclean** [comment]
Doesn't look like [`Listing.vue`](https://github.com/statamic/cms/blob/3.1/resources/js/components/Listing.vue) has been changed that recently (last change was on April 1st). I can probably live with that little bug until I figure out what's causing it. Once deleted, the listing will just reload anyway so it's not a massive problem.

**Review by duncanmcclean — APPROVED** [review]
Thanks Ryan! Everything looks good, I'm planning on merging this tomorrow - just need to write some tests for the new controller first but I'll leave them for tomorrow (it's 11pm haha)

**Comment by ryanmitchell** [comment]
I pushed a quick hack around the data hiding on delete / cancel 

As you can see visibleColumns is somewhere being set to an empty array - I had a long look around the Statamic vue files we’re pulling in and can’t see where its happening. I bet it will be obvious to someone who has worked on that code, so it might be worth reaching out to one of the core team?

**Comment by duncanmcclean** [comment]
That's weird. I'll do some digging later. 

I did notice something change in dev tools when I opened the delete modal but didn't track down what it was 😅

**Comment by duncanmcclean** [comment]
Merged!! Thanks again for all your work, I've got a couple of little things still to do but I'll do them in a separate PR (convert listing buttons to proper actions, get rid of the `listing` config options)

**Comment by ryanmitchell** [comment]
No problem. If there’s anything else I can help with just ask

**Comment by duncanmcclean** [comment]
Released in [v2.1.0](https://github.com/doublethreedigital/runway/releases/tag/v2.1.0)

**Pull request merged**
